### ISSUE TITLE
[Week 4] [Critical] withdraw_from_yield inverts PnL signs against PROGRAM_GUIDE, gains shrink pool.total_sol

### ISSUE THREAD

**Issue by Leihyn** [Frankcastleauditor/Solana-Audit-Arena#144]
## Finding

**Week**: 4
**Researcher**: [Leihyn](https://github.com/Leihyn) + [@faruukku](https://x.com/faruukku)
**Severity**: Critical
**Category**: Logic error / inverted accounting / spec divergence
**Affected function**: `stake_v2::withdraw_from_yield`

## Description

`withdraw_from_yield` applies the strategy PnL to `pool.total_sol` with both signs reversed against the documented behavior. The shipped spec is explicit on what should happen.

`programs/stake_v2/PROGRAM_GUIDE.md:176-177`:
> As `total_sol` grows (positive yield) each `frankSOL` redeems for more SOL.
> As `total_sol` shrinks (loss) each `frankSOL` redeems for less SOL.

`programs/yield_generator/PROGRAM_GUIDE.md:329-335`:
> ```
> vault_delta = vault_after - vault_before
> pnl = vault_delta - principal_returned   (signed)
> ```
>
> In negative mode, `vault_delta < principal_returned` so `pnl < 0` and
> `pool.total_sol` decreases — automatically depressing the `frankSOL` price.

So the spec says: gain → `total_sol` grows; loss → `total_sol` shrinks.

The code does the opposite. `programs/stake_v2/src/instructions/withdraw_from_yield.rs:55-82`:

```rust
let vault_before = ctx.accounts.vault.lamports();
// ... CPI yield_generator::withdraw transfers (principal +/- yield) into vault ...
let vault_after = ctx.accounts.vault.lamports();
let total_return = checked_sub_u64(vault_after, vault_before)?;
let pnl_i128 = (total_return as i128) - (principal_returned as i128);

pool.deployed_sol = PodU64::from(checked_sub_u64(
    pool.deployed_sol.get(),
    principal_returned,
)?);
if pnl_i128 >= 0 {
    let gain = u64::try_from(pnl_i128).map_err(|_| StakeError::MathOverflow)?;
    pool.total_sol = PodU64::from(checked_sub_u64(pool.total_sol.get(), gain)?);
} else {
    let loss = u64::try_from(-pnl_i128).map_err(|_| StakeError::MathOverflow)?;
    pool.total_sol = PodU64::from(checked_add_u64(pool.total_sol.get(), loss)?);
}
```

- gain (`pnl_i128 >= 0`): code subtracts `gain` from `pool.total_sol`. Spec says it should add.
- loss (`pnl_i128 < 0`): code adds `loss` to `pool.total_sol`. Spec says it should subtract.

Both branches are wired backwards. The `deploy_to_yield` side is consistent (it only mutates `deployed_sol`, not `total_sol`, which is correct). The inversion is purely in the settlement step.

## Impact

Two distinct failure modes, both terminal for the protocol.

**1. Honest yield destroys redemption value.**

State: `total_sol = 100 SOL`, `deployed_sol = 100 SOL`, `franksol_supply = 100 SOL`. Strategy returns +5 SOL on a normal cycle. `vault_after - vault_before = 105 SOL`, `principal_returned = 100`, `pnl = +5`.

| | spec says | code does |
|---|---|---|
| `pool.total_sol` after | 105 | 95 |
| `franksol_supply` | 100 | 100 |
| exchange rate `total_sol / supply` | 1.05 (gain captured) | 0.95 (gain inverted) |

Vault physically holds 105 SOL. Pool accounting says only 95 SOL backs the 100 frankSOL supply. The 10 SOL gap is unreachable through any normal entrypoint. After enough cycles, `pool.total_sol` will underflow `checked_sub_u64`, freezing all `withdraw_from_yield` calls and locking the strategy permanently.

**2. Strategy losses inflate redemption value, enabling a bank run.**

`yield_generator::set_yield_direction` (PROGRAM_GUIDE.md:316: *"The authority can toggle `yield_direction_positive` at any time."*) lets the authority flip the strategy into loss mode with no timelock. When `withdraw_from_yield` settles a loss cycle, `pool.total_sol` is increased instead of decreased. The exchange rate jumps. Any holder unstaking right after is paid more SOL than the protocol holds, draining stakers who unstake later.

A controlled adversary path:
- Attacker holds `X frankSOL`.
- Attacker (or someone watching the authority) waits for `yield_generator` authority to flip `yield_direction_positive = false`, or coordinates with the authority.
- `withdraw_from_yield` runs with `total_return < principal_returned`. The handler adds `loss` to `total_sol`.
- Attacker unstakes immediately. `franksol_to_sol(X, total_sol_inflated, supply)` returns more SOL than the share is worth.
- Late stakers unstake later against a vault that has paid out the inflated claim, hitting `InsufficientVaultBalance`.

## Proof of Concept

Numeric trace, gain branch:

| step | total_sol before | deployed_sol | pnl | code does | total_sol after | spec value |
|------|------------------|--------------|-----|-----------|-----------------|------------|
| init | 100 | 0   | -   | -                  | 100 | 100 |
| deploy 100 | 100 | 100 | -   | deployed_sol += 100 | 100 | 100 |
| withdraw, return = 105, pnl = +5 | 100 | 100 | +5 | total_sol -= 5 | **95**  | **105** |
| withdraw, return = 95,  pnl = -5 | 100 | 100 | -5 | total_sol += 5 | **105** | **95**  |

LiteSVM repro (drop into `programs/stake_v2/tests/`; build both `.so` first):

```rust
#[test]
fn withdraw_from_yield_inverts_pnl_signs() {
    use stake_v2::{accounts as s, instruction as si};
    let mut svm = LiteSVM::new();
    svm.add_program(stake_v2::id(), &load("stake_v2")).unwrap();
    svm.add_program(yield_generator::id(), &load("yield_generator")).unwrap();

    let admin = Keypair::new();
    let fm = Keypair::new();
    let user = Keypair::new();
    svm.airdrop(&admin.pubkey(), 10_000_000_000).unwrap();
    svm.airdrop(&fm.pubkey(), 1_000_000_000).unwrap();
    svm.airdrop(&user.pubkey(), 200_000_000_000).unwrap();

    // bootstrap_yield_generator(&mut svm, &admin);
    // bootstrap_stake_v2(&mut svm, &admin, fm.pubkey(), admin.pubkey());
    // user_stakes(&mut svm, &user, 100_000_000_000); // 100 SOL

    let (pool, _) = Pubkey::find_program_address(&[b"pool"], &stake_v2::id());
    let (vault, _) = Pubkey::find_program_address(&[b"vault"], &stake_v2::id());
    let (yield_vault, _) = Pubkey::find_program_address(&[b"yield_vault"], &yield_generator::id());

    let total_sol_before = read_pool_total_sol(&svm, &pool); // 100 SOL

    // fm deploys 100 SOL into yield strategy.
    send_deploy_to_yield(&mut svm, &fm, 100_000_000_000);

    // Simulate the strategy returning +5 SOL by airdropping into yield_vault.
    svm.airdrop(&yield_vault, 5_000_000_000).unwrap();

    // fm withdraws principal=100 SOL with +5 SOL gain.
    send_withdraw_from_yield(&mut svm, &fm, 100_000_000_000, 5_000_000_000);

    let total_sol_after = read_pool_total_sol(&svm, &pool);

    // Spec: total_sol_after == 105 SOL (gain captured).
    // Code: total_sol_after == 95 SOL (gain inverted).
    assert_eq!(total_sol_after, 95_000_000_000, "PnL signs are inverted");
    assert!(total_sol_after < total_sol_before,
            "honest yield reduced pool.total_sol instead of growing it");

    // Vault physically holds 105 SOL of liquid + rent. Pool says 95.
    let vault_balance = svm.get_balance(&vault).unwrap();
    let rent = svm.get_sysvar::<solana_rent::Rent>().minimum_balance(0);
    assert_eq!(vault_balance, rent + 105_000_000_000);
    // Discrepancy of 10 SOL: 5 from the inverted gain + 5 actually earned.
}
```

## Recommended Fix

Swap the two arithmetic ops in `programs/stake_v2/src/instructions/withdraw_from_yield.rs:76-82`:

```rust
if pnl_i128 >= 0 {
    let gain = u64::try_from(pnl_i128).map_err(|_| StakeError::MathOverflow)?;
    pool.total_sol = PodU64::from(checked_add_u64(pool.total_sol.get(), gain)?);
} else {
    let loss = u64::try_from(-pnl_i128).map_err(|_| StakeError::MathOverflow)?;
    pool.total_sol = PodU64::from(checked_sub_u64(pool.total_sol.get(), loss)?);
}
```

After the fix, the code matches PROGRAM_GUIDE.md:177 and 335: positive yield grows `total_sol` and the redemption rate, losses shrink it.

**Comment by novoyd** [comment]
This looks valid to me, but I’d rate it Medium rather than Critical. The PnL direction is definitely inverted, so gains reduce `pool.total_sol` and losses inflate it, which breaks pool accounting. I don’t think it reaches Critical because it is exercised through the fund-manager yield-withdraw path, not by any user through a normal public action.

**Comment by Frankcastleauditor** [comment]
**Verdict: ✅ VALID — High**
**Score: 7 points**

This will be classified as high since it leads to loss of funds and value, does not require any specific condition, and affects all users.
